command/fmt: support formatting multiple files

zimbatm opened the pull request:
All the code infrastructure was there to support formatting multiple
files already.

This makes `terraform fmt` more flexible and also compliant with the
[treefmt formatted spec](https://numtide.github.io/treefmt/docs/formatters-spec.html)

Fixes https://github.com/numtide/treefmt/issues/97

zimbatm commented:
codecov is complaining because the added doc is untested :)

zimbatm commented:
ping?

mloughran commented:
Some more context and justification for this PR from my perspective: treefmt allows dramatically speeding up repository formatting by only attempting to format files that have changed. The current interface to `terraform fmt`, while working great standalone, is too restrictive to support this use-case. This seems like a shame!

zimbatm commented:
rebased to fix the merge conflict

zimbatm commented:
rebased to fix the merge conflicts

flokli commented:
Ping @jbardin. It'd be really nice if this could go in!

zimbatm commented:
/cc @teamterraform

hashicorp-cla commented:
[![CLA assistant check](https://cla.hashicorp.com/pull/badge/not_signed)](https://cla.hashicorp.com/hashicorp/terraform?pullRequest=28191)


Thank you for your submission! We require that all contributors sign our Contributor License Agreement ("CLA") before we can accept the contribution. [Read and sign the agreement](https://cla.hashicorp.com/hashicorp/terraform?pullRequest=28191)

[Learn more about why HashiCorp requires a CLA and what the CLA includes](https://www.hashicorp.com/cla)

<sub>Have you signed the CLA already but the status is still pending? [Recheck it](https://cla.hashicorp.com/check/hashicorp/terraform?pullRequest=28191).</sub>

aldoborrero commented:
This feature is useful to have!

mloughran commented:
Some more justification for this PR: Currently the only way to run terraform via `treefmt` is it to use a wrapper script (e.g. see https://github.com/numtide/treefmt/blob/master/treefmt.toml#L77)

Unfortunately, the startup speed of `terraform fmt` means that running it separately for each file is very slow. On my project (64 terraform files) this is already 23s. Output from `treefmt`:

	[INFO]: #terraform: 64 files processed in 23.24s

Whereas

	terraform fmt -recursive  0.48s user 0.90s system 83% cpu 1.659 total

This this is just about acceptable for local use cases (hopefully the files don't change much, and treefmt only runs on changed files), but it's a deal-breaker for use cases such as formatting in CI.

crw commented:
Hi @zimbatm, apologies for the late reply. We should be able to review this within a few weeks. Thanks for your patience.

alisdair reviewed (changes requested):
Thanks for this contribution! This looks like a nice improvement. There are a couple of things to fix before we can merge:

- There's a merge conflict due to some recent help text changes. Can you rebase and fix that?
- We need to update the website docs to note the new functionality. Can you update [the docs file here](https://github.com/hashicorp/terraform/blob/main/website/docs/cli/commands/fmt.mdx) to describe how `terraform fmt` works after your changes?

zimbatm commented:
rebased!
* The code didn't change
* I re-did the doc, copying some of the text from the website
* Renamed `PATH...` to `[target...]` to better follow the other commands.
* Updated the cmd and website descriptions to be more in sync.

alisdair reviewed (approved):
Thanks, this is great!
